File: backend/apps/api/glean_api/routers/entries.py
"""
Entries router.

Provides endpoints for reading and managing feed entries.
"""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status

from glean_core.schemas import (
    EntryListResponse,
    EntryResponse,
    UpdateEntryStateRequest,
    UserResponse,
)
from glean_core.services import EntryService

from ..dependencies import get_current_user, get_entry_service

router = APIRouter()


@router.get("")
async def list_entries(
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    entry_service: Annotated[EntryService, Depends(get_entry_service)],
    feed_id: str | None = None,
    is_read: bool | None = None,
    is_liked: bool | None = None,
    read_later: bool | None = None,
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
) -> EntryListResponse:
    """
    Get entries with filtering and pagination.

    Args:
        current_user: Current authenticated user.
        entry_service: Entry service.
        feed_id: Optional filter by feed ID.
        is_read: Optional filter by read status.
        is_liked: Optional filter by liked status.
        read_later: Optional filter by read later status.
        page: Page number (1-indexed).
        per_page: Items per page (max 100).

    Returns:
        Paginated list of entries.
    """
    return await entry_service.get_entries(
        user_id=current_user.id,
        feed_id=feed_id,
        is_read=is_read,
        is_liked=is_liked,
        read_later=read_later,
        page=page,
        per_page=per_page,
    )


@router.get("/{entry_id}")
async def get_entry(
    entry_id: str,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    entry_service: Annotated[EntryService, Depends(get_entry_service)],
) -> EntryResponse:
    """
    Get a specific entry.

    Args:
        entry_id: Entry identifier.
        current_user: Current authenticated user.
        entry_service: Entry service.

    Returns:
        Entry details.

    Raises:
        HTTPException: If entry not found or user not subscribed to feed.
    """
    try:
        return await entry_service.get_entry(entry_id, current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from None


@router.patch("/{entry_id}")
async def update_entry_state(
    entry_id: str,
    data: UpdateEntryStateRequest,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    entry_service: Annotated[EntryService, Depends(get_entry_service)],
) -> EntryResponse:
    """
    Update entry state (read, liked, read later).

    Args:
        entry_id: Entry identifier.
        data: State update data.
        current_user: Current authenticated user.
        entry_service: Entry service.

    Returns:
        Updated entry.

    Raises:
        HTTPException: If entry not found.
    """
    try:
        return await entry_service.update_entry_state(entry_id, current_user.id, data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from None


@router.post("/mark-all-read")
async def mark_all_read(
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    entry_service: Annotated[EntryService, Depends(get_entry_service)],
    feed_id: str | None = None,
) -> dict[str, str]:
    """
    Mark all entries as read.

    Args:
        current_user: Current authenticated user.
        entry_service: Entry service.
        feed_id: Optional filter by feed ID.

    Returns:
        Success message.
    """
    await entry_service.mark_all_read(current_user.id, feed_id)
    return {"message": "All entries marked as read"}

File: frontend/packages/api-client/src/services/entries.ts
import type {
  EntryWithState,
  EntryListResponse,
  ParagraphTranslationsResponse,
  TranslateTextsResponse,
  TranslationResponse,
  UpdateEntryStateRequest,
} from '@glean/types'
import { ApiClient } from '../client'

/**
 * Entries API service.
 *
 * Handles entry listing, retrieval, and state management.
 */
export class EntryService {
  constructor(private client: ApiClient) {}

  /**
   * Get entries with filtering and pagination.
   */
  async getEntries(params?: {
    feed_id?: string
    folder_id?: string
    is_read?: boolean
    read_later?: boolean
    collected_after?: string
    collected_before?: string
    page?: number
    per_page?: number
    view?: 'timeline' | 'today-board'
  }, options?: { signal?: AbortSignal }): Promise<EntryListResponse> {
    return this.client.get<EntryListResponse>('/entries', { params, signal: options?.signal })
  }

  /**
   * Get entries collected during a server-local day.
   *
   * This endpoint intentionally returns a bounded aggregate instead of timeline pages.
   */
  async getTodayEntries(params: {
    date?: string
    feed_id?: string
    folder_id?: string
    limit?: number
  }, options?: { signal?: AbortSignal }): Promise<EntryListResponse> {
    return this.client.get<EntryListResponse>('/entries/today', {
      params,
      signal: options?.signal,
    })
  }

  /**
   * Get a specific entry.
   */
  async getEntry(entryId: string, options?: { signal?: AbortSignal }): Promise<EntryWithState> {
    return this.client.get<EntryWithState>(`/entries/${entryId}`, { signal: options?.signal })
  }

  /**
   * Update entry state (read and read later).
   */
  async updateEntryState(entryId: string, data: UpdateEntryStateRequest): Promise<EntryWithState> {
    return this.client.patch<EntryWithState>(`/entries/${entryId}`, data)
  }

  /**
   * Mark all entries as read.
   */
  async markAllRead(feedId?: string, folderId?: string): Promise<{ message: string }> {
    return this.client.post<{ message: string }>('/entries/mark-all-read', {
      feed_id: feedId,
      folder_id: folderId,
    })
  }

  // Translation endpoints

  /**
   * Request translation of an entry.
   * If targetLanguage is not provided, auto-detects source and picks the opposite.
   */
  async translateEntry(
    entryId: string,
    targetLanguage?: string | null
  ): Promise<TranslationResponse> {
    return this.client.post<TranslationResponse>(`/entries/${entryId}/translate`, {
      target_language: targetLanguage ?? null,
    })
  }

  /**
   * Get cached translation for an entry.
   */
  async getTranslation(entryId: string, targetLanguage: string): Promise<TranslationResponse> {
    return this.client.get<TranslationResponse>(`/entries/${entryId}/translation/${targetLanguage}`)
  }

  /**
   * Translate an array of text strings synchronously.
   * Used for viewport-based sentence-level translation.
   */
  async translateTexts(
    texts: string[],
    targetLanguage: string,
    sourceLanguage: string = 'auto',
    entryId?: string,
  ): Promise<TranslateTextsResponse> {
    return this.client.post<TranslateTextsResponse>('/entries/translate-texts', {
      texts,
      target_language: targetLanguage,
      source_language: sourceLanguage,
      entry_id: entryId,
    })
  }

  /**
   * Get persisted paragraph-level translations for an entry.
   */
  async getParagraphTranslations(
    entryId: string,
    targetLanguage: string,
  ): Promise<ParagraphTranslationsResponse> {
    return this.client.get<ParagraphTranslationsResponse>(
      `/entries/${entryId}/paragraph-translations`,
      { params: { target_language: targetLanguage } },
    )
  }
}
